swarm_dial.go

```diff
@@ -359,6 +359,33 @@ func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.
 
 	var active int
 	for {
+		// Check for context cancellations and/or responses first.
+		select {
+		case <-ctx.Done():
+			if exitErr == defaultDialFail {
+				exitErr = ctx.Err()
+			}
+			return nil, exitErr
+		case resp := <-respch:
+			active--
+			if resp.Err != nil {
+				log.Infof("got error on dial to %s: %s", resp.Addr, resp.Err)
+				// Errors are normal, lots of dials will fail
+				exitErr = resp.Err
+
+				if remoteAddrs == nil && active == 0 {
+					return nil, exitErr
+				}
+			} else if resp.Conn != nil {
+				return resp.Conn, nil
+			}
+
+			// We got a result, try again from the top.
+			continue
+		default:
```

Comment:
This is a bunch of duplicated code. Not sure if there's a better way.

Reply:
The context cancellation check is crucial. Draining responses first before starting more dials is an optimisation, but a welcome one.